<launch>
  <!-- Single drone offboard control for visPlanner - SITL version
       Supports configuration files like Elastic-Tracker -->
  
  <arg name="drone_id" default="0"/>
  <arg name="drone_ns" default="drone0"/>
  
  <!-- Configuration file path (optional, can use default parameters) -->
  <arg name="config_file" default="" doc="Path to YAML configuration file (optional)"/>
  
  <!-- Default parameters (can be overridden by config_file) -->
  <arg name="takeoff_alt" default="1.2"/>
  <arg name="takeoff_time" default="10.0"/>
  <arg name="landing_time" default="2.0"/>
  <arg name="landing_max_vel" default="0.3"/>
  <arg name="goto_max_vel" default="1.0"/>
  <arg name="goto_tol" default="0.05"/>
  <arg name="end_traj_wait_time" default="5.0"/>
  <arg name="traj_use_position_control" default="true" doc="TRAJ control mode: false=Rate Control, true=Position Control"/>
  
  <!-- Optional GOTO target (NAN means no target, will hover after takeoff) -->
  <arg name="goto_x" default="NAN"/>
  <arg name="goto_y" default="NAN"/>
  <arg name="goto_z" default="NAN"/>
  
  <!-- Formation parameters (if using multiple drones) -->
  <arg name="num_drones" default="1"/>
  <arg name="inward_offset" default="0.0"/>
  <arg name="payload_offset_x" default="0.0"/>
  <arg name="payload_offset_y" default="0.0"/>
  
  <!-- Set use_sim_time to false (use system time) -->
  <param name="/use_sim_time" value="false"/>
  
  <!-- Load parameters from config_file if provided (loads into node's private namespace) -->
  <group if="$(eval config_file != '')">
    <rosparam file="$(arg config_file)" ns="offboard_fsm_drone_$(arg drone_id)" command="load"/>
  </group>
  
  <!-- Offboard FSM Node -->
  <node name="offboard_fsm_drone_$(arg drone_id)" 
        pkg="offboard_state_machine" 
        type="offboard_fsm_node" 
        output="screen">
    
    <!-- Remap MAVROS topics to correct namespace (for SITL mode with drone0 namespace) -->
    <remap from="/mavros/state" to="/$(arg drone_ns)/mavros/state"/>
    <remap from="/mavros/local_position/pose" to="/$(arg drone_ns)/mavros/local_position/pose"/>
    <remap from="/mavros/local_position/velocity_local" to="/$(arg drone_ns)/mavros/local_position/velocity_local"/>
    <remap from="/mavros/cmd/arming" to="/$(arg drone_ns)/mavros/cmd/arming"/>
    <remap from="/mavros/set_mode" to="/$(arg drone_ns)/mavros/set_mode"/>
    <remap from="/mavros/setpoint_raw/local" to="/$(arg drone_ns)/mavros/setpoint_raw/local"/>
    
    <param name="drone_id" value="$(arg drone_id)"/>
    <param name="takeoff_alt" value="$(arg takeoff_alt)"/>
    <param name="takeoff_time" value="$(arg takeoff_time)"/>
    <param name="landing_time" value="$(arg landing_time)"/>
    <param name="landing_max_vel" value="$(arg landing_max_vel)"/>
    <param name="goto_max_vel" value="$(arg goto_max_vel)"/>
    <param name="goto_tol" value="$(arg goto_tol)"/>
    <param name="end_traj_wait_time" value="$(arg end_traj_wait_time)"/>
    <param name="traj_use_position_control" value="$(arg traj_use_position_control)"/>
    
    <param name="goto_x" value="$(arg goto_x)"/>
    <param name="goto_y" value="$(arg goto_y)"/>
    <param name="goto_z" value="$(arg goto_z)"/>
    
    <param name="num_drones" value="$(arg num_drones)"/>
    <param name="inward_offset" value="$(arg inward_offset)"/>
    <param name="payload_offset_x" value="$(arg payload_offset_x)"/>
    <param name="payload_offset_y" value="$(arg payload_offset_y)"/>
    
    <param name="timer_period" value="0.02"/>
    <param name="alt_tol" value="0.01"/>
    <param name="goto_accel_time" value="3.0"/>
  </node>
  
</launch>

